### PR TITLE
Level 3

### PR DESCRIPTION
**Pull request by DenyaAbramenkov**
Done level_3 with documentation and tests

### CODE REVIEW

File: LibraryApi/Controllers/AuthorsController.cs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+using Microsoft.AspNetCore.Mvc;
+using LibraryApi.Models;
+using LibraryApi.Services;
+using System.Collections.Generic;
+
+namespace LibraryApi.Controllers
+{
+    [Route("api/authors")]
+    [ApiController]
+    public class AuthorsController : ControllerBase
```
Comment (anton-borodulin-softserve) [inline LibraryApi/Controllers/AuthorsController.cs:14]: Add documentation here as-well.

File: LibraryApi/Controllers/BooksController.cs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+using System.Collections.Generic;
+
+namespace LibraryApi.Controllers
+{
+    [Route("api/books")]
+    [ApiController]
+    public class BooksController : ControllerBase
+    {
+        /// <summary>
+        /// Library Service
```
Comment (anton-borodulin-softserve) [inline LibraryApi/Controllers/BooksController.cs]: It`s a good style to put dot in the end of any documentation.

File: LibraryServiceTest/AuthorServiceTest.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+using System.Collections.Generic;
+using Moq;
+using System.Linq;
+
+namespace LibraryServiceTest
+{
+    [TestClass]
+    public class AuthorServiceTest
+    {
+
```
Comment (anton-borodulin-softserve) [inline LibraryServiceTest/AuthorServiceTest.cs]: redundant line

File: LibraryServiceTest/AuthorServiceTest.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        private static Mock<IDataProvider> _data;
```
Comment (anton-borodulin-softserve) [inline LibraryServiceTest/AuthorServiceTest.cs:13]: Why static?

File: LibraryServiceTest/AuthorServiceTest.cs [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+                new Author("The Gratest", "A", 1984),
+                new Author("Good", "B", 1978)
+            };
+            _data = new MockDataImitation().MockSetAuthors(_authors);
+        }
+
+        [TestMethod]
+        public void GetAllAuthorsTest()
+        {
+            var library = _data.Object;
```
Comment (anton-borodulin-softserve) [inline LibraryServiceTest/AuthorServiceTest.cs:31]: Use AAA principle in all unit tests

File: LibraryServiceTest/AuthorServiceTest.cs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+            List<Author> result = new Library(library).GetAllAuthors().ToList();
+
+            CollectionAssert.AreEqual(_authors, result);
+        }
+
+        [TestMethod]
+        public void GetAuthorByIdTest()
```
Comment (anton-borodulin-softserve) [inline LibraryServiceTest/AuthorServiceTest.cs]: I'm not a fun for adding ending 'Test' it doesn't take any additional sense. As you will expect that when you are running tests, tests will run. I prefer AuthorService_GetAuthorById name. Because when you will run the test from test explorer you will understand in a second where and what failed. Or AuthorService_GetAuthorById_Positive / AuthorService_GetAuthorById_Null_Ref_Exception_Should_Throw. etc. It`s allowing to know more even if you are looking only on test explorer.

File: LibraryServiceTest/MockDataImitation.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+using LibraryApi.Models;
+using LibraryApi.Services;
+
+namespace LibraryServiceTest
+{
+    class MockDataImitation: Mock<IDataProvider>
+    {
+        public MockDataImitation MockSetAuthors(List<Author> authors)
+        {
+            Setup(library => library.SetAuthors()).Returns(authors);
```
Comment (anton-borodulin-softserve) [inline LibraryServiceTest/MockDataImitation.cs:14]: If you have not over complicated mock setup I prefer to make setup in each test. It allows you to see what data were set and what to expect.

File: LibraryServices/Models/Author.cs [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+        /// <summary>
+        /// Gets the Author's Id
+        /// </summary>
+        public int Id { get; private set; }
+
+        /// <summary>
+        /// Gets or sets the Author's Name
+        /// </summary>
+        [Required(ErrorMessage = "Author's Name is required")]
```
Comment (anton-borodulin-softserve) [inline LibraryServices/Models/Author.cs]: Do you use anywhere this error message? If not maybe you should if you wrote it.

File: LibraryServices/Models/Author.cs [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        /// Gets or sets the Author's Surname
+        /// </summary>
+        public string Surname { get; set; }
+
+        /// <summary>
+        /// Gets or sets the Author's Year of Birthday
+        /// </summary>
+        public int YearOfBirthday { get; set; }
+
+        public override bool Equals(object author)
```
Comment (anton-borodulin-softserve) [inline LibraryServices/Models/Author.cs:55]: Documentation is missing

File: LibraryServices/Models/Author.cs [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        {
+            Author newauthor = author as Author;
```
Comment (anton-borodulin-softserve) [inline LibraryServices/Models/Author.cs]: newAuthor

File: LibraryServices/Models/Genre.cs [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+        /// <summary>
+        /// Gets or sets the Genres's Name
+        /// </summary>
+        [Required(ErrorMessage = "Genre's Name is required")]
+        public string Name { get; set; }
+
+        public override bool Equals(object genre)
+        {
+            Genre newgenre = genre as Genre;
```
Comment (anton-borodulin-softserve) [inline LibraryServices/Models/Genre.cs]: newGenre

File: LibraryServices/Services/DataProvider.cs [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+            List<Genre> genres = new List<Genre>
+            {
+                new Genre("Comedy"),
+                new Genre("Tragedy"),
+                new Genre("Detective")
+            };
+
+            foreach (Genre genre in genres)
+            {
+                yield return genre;
```
Comment (anton-borodulin-softserve) [inline LibraryServices/Services/DataProvider.cs]: why not - return genres; ?

File: LibraryServices/Services/DataProvider.cs [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+        {
+            IEnumerable<KeyValuePair<int, int>> booksAndGenreConncetion = new List<KeyValuePair<int, int>>
+            {
+                new KeyValuePair<int, int>(1, 1),
+                new KeyValuePair<int, int>(1, 2),
+                new KeyValuePair<int, int>(1, 3),
+                new KeyValuePair<int, int>(2, 2)
+            };
+            
+            foreach(KeyValuePair<int, int> keyValue in booksAndGenreConncetion)
```
Comment (anton-borodulin-softserve) [inline LibraryServices/Services/DataProvider.cs]: why are you returning the result in the loop instead of simple return booksAndGenreConncetion;

File: LibraryServices/Services/ILibrary.cs [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+        /// </summary>
+        /// <param name="author_Id">The author identifier.</param>
+        /// <returns>Enumeration of books.</returns>
+        IEnumerable<Book> GetAuthorBooks(int authorId);
+        #endregion
+
+        /// <summary>
+        /// Region of Genre's Services
+        /// </summary>
+        #region Janre
```
Comment (anton-borodulin-softserve) [inline LibraryServices/Services/ILibrary.cs]: Genre

File: LibraryServices/Services/ILibrary.cs [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+
+        /// <summary>
+        /// Gets all books of supposed genre.
+        /// </summary>
+        /// <param name="genre_Id">The genre identifier.</param>
+        /// <returns>Enumeration of books.</returns>
+        IEnumerable<Book> GetAllGenreBooks(int genreId);
+
+        #endregion
+
```
Comment (anton-borodulin-softserve) [inline LibraryServices/Services/ILibrary.cs:161]: remove all redundant lines.

File: LibraryServices/Services/ILibrary.cs [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+        /// <summary>
+        /// Update Genre
+        /// </summary>
+        /// <param name="id">Genre's Id</param>
+        /// <param name="genre">New Genre's Id</param>
+        /// <returns>Updated Genre</returns>
+        Genre UpdateGenre(int id, Genre genre);
+
+        /// <summary>
+        /// Delete Genre from collection
```
Comment (anton-borodulin-softserve) [inline LibraryServices/Services/ILibrary.cs]: here you don't have dot

File: LibraryServices/Services/ILibrary.cs [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+        /// </summary>
+        /// <param name="id">Genre's Id</param>
+        void DeleteGenre(int id);
+
+        /// <summary>
+        /// Gets all books of supposed genre.
```
Comment (anton-borodulin-softserve) [inline LibraryServices/Services/ILibrary.cs:154]: but here you have. It's okay to have them or not. But not the mixed option.

File: LibraryServices/Services/ILibrary.cs [hunk 18]
```diff
@@ -0,0 +1,172 @@
+﻿using LibraryApi.Models;
+using System.Collections.Generic;
+
+
+namespace LibraryApi.Services
+{
+    public interface ILibrary
```
Comment (anton-borodulin-softserve) [inline LibraryServices/Services/ILibrary.cs:7]: It became too big. It`s time to split it into three services.

File: LibraryServices/Services/Library.cs [hunk 19]
```diff
@@ -0,0 +1,300 @@
...
+        /// <summary>
+        /// Update Book in collection
+        /// </summary>
+        /// <param name="id">Book's Id</param>
+        /// <param name="book">New Book info</param>
+        /// <returns>Updated Book</returns>
+        public Book UpdateBook(int id, Book book)
+        {
+            Book bookToUpdate = _booksInLibrary.Find((Book) => Book.Id == id);
+            if (book != null)
```
Comment (anton-borodulin-softserve) [inline LibraryServices/Services/Library.cs:96]: hm if book is null. Probably it's incorrect data for update. Maybe it's a nice idea to throw an exception, handle it and return message to the client. To avoid misunderstanding, the client can think that the update was successful.

File: LibraryServices/Services/Library.cs [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+        #endregion
+
+        #region AuthorRegion
+
+        /// <summary>
+        /// Return All Authors
+        /// </summary>
+        public IEnumerable<Author> GetAllAuthors()
+        {
+            foreach (Author author in _authors)
```
Comment (anton-borodulin-softserve) [inline LibraryServices/Services/Library.cs]: I'm starting to think that I don't know something. And you have some big reason to write 3 lines instead of a simple one with the return.

File: LibraryServices/Services/Library.cs [hunk 21]
```diff
@@ -0,0 +1,300 @@
...
+
+        /// <summary>
+        /// Update Author
+        /// </summary>
+        /// <param name="id">Author's Id</param>
+        /// <param name="author">New Author</
+        public Author UpdateAuthor(int id, Author author)
+        {
+            Author authorToUpdate = _authors.Find((Author) => Author.Id == id);
+            if (authorToUpdate != null)
```
Comment (anton-borodulin-softserve) [inline LibraryServices/Services/Library.cs:178]: I suppose you are not handling the case when authorToUpdate is null. And not using the result from this method.

File: LibraryServices/Services/Library.cs (hunk 19 above)
Comment (anton-borodulin-softserve) [reply LibraryServices/Services/Library.cs:96]: Also, you should add a unit tests for such negative scenario. As I saw you tested only positive cases

File: LibraryServices/Services/Library.cs [hunk 22]
```diff
@@ -0,0 +1,300 @@
...
+                    _booksInLibrary.Remove(book);
+                }
+            }
+            _authors.Remove(authorToDelete);
+        }
+
+
+        public IEnumerable<Book> GetAuthorBooks(int author_Id)
+        {
+            foreach (var authorBook in _booksInLibrary.Where(Book => Book.AuthorId == author_Id))
```
Comment (anton-borodulin-softserve) [inline LibraryServices/Services/Library.cs]: return _booksInLibrary.Where(Book => Book.AuthorId == author_Id);

File: LibraryServices/Services/Library.cs [hunk 23]
```diff
@@ -0,0 +1,300 @@
...
+
+        /// <summary>
+        /// Update Genre
+        /// </summary>
+        /// <param name="id">Genre's Id</param>
+        /// <param name="genre">New Genre's Id</param>
+        /// <returns>Updated Genre</returns>
+        public Genre UpdateGenre(int id, Genre genre)
+        {
+            Genre genreToUpdate = _genres.Find((Genre) => Genre.Id == id);
```
Comment (anton-borodulin-softserve) [inline LibraryServices/Services/Library.cs]: Genre is a local variable inside anonymous method scope. so it should start from a lower letter. - genre

File: LibraryServices/Services/Library.cs [hunk 24]
```diff
@@ -0,0 +1,300 @@
...
+        /// <param name="id">Genre's Id</param>
+        public void DeleteGenre(int id)
+        {
+            Genre genreToDelete = _genres.Find(Genre => Genre.Id == id);
+            _genres.Remove(genreToDelete);
+        }
+
+        public IEnumerable<Book> GetAllGenreBooks(int genre_Id)
+        {
+            foreach (var genreBook in _bookAndGenresConnection.Where(bookGenre => bookGenre.Value == genre_Id))
```
Comment (anton-borodulin-softserve) [inline LibraryServices/Services/Library.cs]: return _bookAndGenresConnection.Where(bookGenre => bookGenre.Value == genre_Id);

File: LibraryServices/Services/Library.cs [hunk 25]
```diff
@@ -0,0 +1,300 @@
...
+            {
+                yield return GetBook(genreBook.Key);
+            }
+        }
+
+        #endregion
+
+    }
+
```
Comment (anton-borodulin-softserve) [inline LibraryServices/Services/Library.cs]: redundant lines

File: LibraryServices/Services/Library.cs [hunk 26]
```diff
@@ -0,0 +1,300 @@
...
+        public IEnumerable<Book> GetAllGenreBooks(int genre_Id)
+        {
+            foreach (var genreBook in _bookAndGenresConnection.Where(bookGenre => bookGenre.Value == genre_Id))
+            {
+                yield return GetBook(genreBook.Key);
+            }
+        }
+
+        #endregion
+
```
Comment (anton-borodulin-softserve) [inline LibraryServices/Services/Library.cs:294]: and here

**Review by anton-borodulin-softserve — APPROVED** [review]
There are alot of comments.
 But I like how you are working after the last demo. I see great progress. Keep it up.

File: LibraryServiceTest/BookServiceTest.cs [hunk 27]
```diff
@@ -0,0 +1,87 @@
...
+using System.Collections.Generic;
+using Moq;
+using System.Linq;
+
+namespace LibraryServiceTest
+{
+    [TestClass]
+    public class BookServiceTest
+    {
+
```
Comment (Kropisha) [inline LibraryServiceTest/BookServiceTest.cs]: Стайл коп не ругает за пустую строку?

File: LibraryServiceTest/GenreServiceTest.cs [hunk 28]
```diff
@@ -0,0 +1,87 @@
...
+            var dataForLib = _data.Object;
+            Library lib = new Library(dataForLib);
+
+            lib.DeleteGenre(1);
+            List<Genre> result = lib.GetAllGenres().ToList();
+            _genres.Remove(_genres[0]);
+
+            CollectionAssert.AreEqual(_genres, result);
+        }
+
```
Comment (Kropisha) [inline LibraryServiceTest/GenreServiceTest.cs:99]: И здесь
